docs/module-sources.rst

```diff
@@ -0,0 +1,562 @@
+          dest-filename: foo.deb
+          only-arches: [aarch64]
+
+``dest`` is used to indicate a directory inside the source directory
+where the given source will be placed:
+
+.. code-block:: yaml
+
+  modules:
+    - name: coolapp
+      buildsystem: simple
+      build-commands:
+        - ...
+      sources:
+        - type: archive
+          url: https://example.org/foo.tar.xz
+          sha256: 834b06e423d360c97197e7abec99b623fdc5ed3a0c39b88d6467e499074585e1
+        - type: git
+          url: https://github.com/foo/bar.git
+          commit: 3f9389edc6cdf3f78a6896d550c236860aed62b2
+          dest: 'src/deps/bar'
+
+The contents of ``foo.tar.xz`` will be placed under
+``.flatpak-builder/build/coolapp/`` and the contents of the ``bar.git``
+repo will go under ``.flatpak-builder/build/coolapp/src/deps/bar``.
+
+This is often useful to replace bundled dependencies inside sources
+with custom versions.
+
+Archive sources
+---------------
+
+These are sources of ``type: archive``. They are downloaded to the
+download directory and extracted into the build directory when the build
+starts.
+
+The download directoy is ``.flatpak-builder/downloads/<checksum of source>/``.
+
+A typical example of using this is:
+
+.. code-block:: yaml
+
+  modules:
+    - name: foo
+      buildsystem: meson
+      sources:
+        - type: archive
+          url: https://example.org/foo-1.0.tar.xz
+          mirror-urls:
+            - https://mirror.example.com/foo-1.0.tar.xz
+          sha256: 26c9007c655483438d83f17070190fc4b6bccb8ff5efc30fcedbb7d162681a7f
+          dest-filename: foo.tar.xz
+          strip-components: 2
+
+- ``mirror-urls`` is a list of alternative urls that will be used for
+  downloading, if the main ``url`` fails.
+
+- The ``sha256`` is the sha256 checksum of the downloaded file, supplied
+  by upstream of the source or calculated using ``sha256sum``.
+
+- ``dest-filename`` is not usually used for this source. It is used to
+  rename the downloaded file. This can be useful if the buildsystem
+  expects a certain name for the archive.
+
+- ``strip-components`` is the number of pathname components to strip
+  during extraction. This defaults to ``1`` if not specified.
+
+.. code-block::
+
+  # orginal archive
+  .
+  └── a
+      ├── b
+      │   ├── c
+      │   │   └── FILE
+      │   └── FILE
+      └── FILE
+
+  # strip-components: 1
+  .
+  ├── b
+  │   ├── c
+  │   │   └── FILE
+  │   └── FILE
+  ├── FILE
+
+
+  # strip-components: 2
+  .
+  ├── c
+  │   └── FILE
+  ├── FILE
+
+
+``flatpak-builder`` supports
+the following archive types: ``"rpm", "tar", "tar-gzip", "tar-compress", "tar-bzip2", "tar-lzip", "tar-lzma", "tar-lzop", "tar-xz", "tar-zst", "zip", "7z"``
+They type of the archive is `gussed` from the suffix of the file basename.
+
+``rpm`` is extracted with ``rpm2cpio``, ``zip`` is extracted with ``unzip``
+and ``7z`` is extratced with ``7z``. The rest are extracted with
+``tar``.
```

Comment:
It should be mentioned the utilities should be installed on the host for flatpak-builder to be able to use them. 7z usually should be discouraged.